tools/c7n_oci/c7n_oci/session.py

```diff
@@ -76,12 +81,19 @@ def __call__(self):
 class Session:
     def __init__(self, config):
         self._config = config
+        self.signer = None
 
-    def client(self, client_string):
+    def client(self, client_string, **kwargs):
         service_name, client_name = client_string.rsplit(".", 1)
         service_module = importlib.import_module(service_name)
         client_class = getattr(service_module, client_name)
-        client_args = {"config": self._config}
+        client_args = {"config": self._config, **kwargs}
+        if os.environ.get('OCI_CLI_AUTH') == 'instance_principal':
```

Comment:
is there a way to detect this?, via metadata server presence with a short timeout, it feels like we're putting a lot of configuration for they provider into environment variables.

Reply:
It is possible to design something like this, however setting up an auto-detect like this will deprive users of the ability to not use instance auth while running custodian on an oci instance.

Reply:
there isn't a standard auth look up in the sdk? ie. environment variables, config files, metadata server? every app needs to reimplement chaining with differently? that's quite odd.

let me rephrase what does the oci cli do?

Reply:
The OCI_CLI_AUTH env. var is used to control the method of authentication for the cli. It is not necessary for us to use the same methodology, I only did that to re-use what the cli is doing. The instance principal auth can be determined easily by querying the metadata url, for e.g. a GET to http://169.254.169.254/opc/v1/instance/region with a 1s timeout. The issue is that I want the users to have a means to not use instance auth when running custodian on an oci instance. One other way to achieve this is to auto detect instance auth and only override it if the user explicitly gives command line arg specifying a config file. Let me know what you think.

Reply:
reusing the same logic as the standard cli sounds good.